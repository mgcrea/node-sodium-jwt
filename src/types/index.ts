import type { RegisteredClaims, PublicClaims, CustomClaims } from "./jwt";

export type Header = {
  alg: "HS512T";
  typ: "JWT";
};

// eslint-disable-next-line @typescript-eslint/ban-types
export type Simplify<T> = { [KeyType in keyof T]: T[KeyType] } & {};

export type Payload = Simplify<RegisteredClaims & PublicClaims & CustomClaims>;
