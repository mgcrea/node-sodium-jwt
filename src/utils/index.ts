export * from './base64';
export * from './casting';
export * from './error';
