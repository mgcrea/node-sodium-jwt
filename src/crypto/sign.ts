import sodium from 'sodium-native';
import {Header, Payload} from '../types';
import {asBuffer, base64UrlEncode} from '../utils';

export const signSync = (payload: Payload, secretKey: Buffer | string, extraHeader: Partial<Header> = {}): string => {
  const header: Header = {
    alg: 'HS512T',
    typ: 'JWT',
    ...extraHeader,
  };
  const output = Buffer.allocUnsafe(sodium.crypto_auth_BYTES);
  const message = `${base64UrlEncode(header)}.${base64UrlEncode(payload)}`;
  sodium.crypto_auth(output, Buffer.from(message), asBuffer(secretKey, 'base64'));
  return `${message}.${output.toString('base64url')}`;
};
