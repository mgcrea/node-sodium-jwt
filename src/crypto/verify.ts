import sodium from "sodium-native";
import type { Header, Payload } from "../types";
import { asArray, asBuffer, createError } from "../utils";

export const verifySync = (
  message: string,
  secretKeys: (string | Buffer) | (string | Buffer)[],
): { header: Header; payload: Payload; rotated: boolean } => {
  const splits = message.split(".");
  if (splits.length !== 3) {
    throw createError("MalformedMessageError", "The message is malformed");
  }
  const [rawHeader, rawPayload, rawSignature] = splits as [string, string, string];

  const header = JSON.parse(Buffer.from(rawHeader, "base64url").toString());
  if (header.alg !== "HS512T") {
    throw createError("UnsupportedAlgorithmError", `The alg "${header.alg}" is unsupported`);
  }

  const signature = Buffer.from(rawSignature, "base64url");
  if (signature.length !== sodium.crypto_auth_BYTES) {
    throw createError("SignatureLengthError", "The signature does not have the required length");
  }

  const cleartext = Buffer.from(message.slice(0, -rawSignature.length - 1), "utf8");

  let rotated = false;
  const success = asArray(secretKeys).some((secretKey, index) => {
    const verified = sodium.crypto_auth_verify(signature, cleartext, asBuffer(secretKey, "base64"));
    rotated = verified && index > 0;
    return verified;
  });

  if (!success) {
    throw createError("VerifyError", "Unable to verify");
  }

  const payload = JSON.parse(Buffer.from(rawPayload, "base64url").toString());

  return { header, payload, rotated };
};
