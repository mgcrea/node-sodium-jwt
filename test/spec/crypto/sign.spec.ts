import { signSync } from "src/crypto";
import type { JWTPayload } from "src/types";
import { describe, expect, it } from "vitest";

const payload: JWTPayload = {
  sub: 7,
  iat: new Date("2022-01-03T16:37:48.612Z").getTime(),
};
const secretKey = "0DfgJOaVrb2quroZavPLp7KJm+hTCBN6hZKnsKQSM+o=";

describe("Sign", () => {
  it("should properly sign a JWT", async () => {
    const jwt = signSync(payload, Buffer.from(secretKey, "base64"));
    expect(jwt).toBeDefined();
    expect(typeof jwt).toBe("string");
    expect(jwt.split(".").length).toBe(3);
    expect(jwt).toMatchSnapshot();
  });
  it("should properly support a stringified secretKey", async () => {
    const jwt = signSync(payload, secretKey);
    expect(jwt).toBeDefined();
    expect(typeof jwt).toBe("string");
    expect(jwt.split(".").length).toBe(3);
    expect(jwt).toMatchSnapshot();
  });
});
