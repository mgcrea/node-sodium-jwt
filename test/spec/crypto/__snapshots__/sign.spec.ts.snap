// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Sign > should properly sign a JWT 1`] = `"eyJhbGciOiJIUzUxMlQiLCJ0eXAiOiJKV1QifQ.eyJzdWIiOjcsImlhdCI6MTY0MTIyNzg2ODYxMn0.a3_bFmeD5vugh3p998QXTdPbAuAJFQzCm-2MfmRVEAg"`;

exports[`Sign > should properly support a stringified secretKey 1`] = `"eyJhbGciOiJIUzUxMlQiLCJ0eXAiOiJKV1QifQ.eyJzdWIiOjcsImlhdCI6MTY0MTIyNzg2ODYxMn0.a3_bFmeD5vugh3p998QXTdPbAuAJFQzCm-2MfmRVEAg"`;
